Billing History: Group domain-related purchases of the same domain

Copons opened the pull request:
Fix #2525 
Require D10703-code

Additional context: p2MSmN-3nc-p2

When a user purchases a domain registration, the receipt shows 2 or 3 different products ("Domain Registration", "Domain Mapping", and sometimes "Private Registration") causing confusion to the user.

This patch groups all the domain-related products for the same domain into one single product, and list them in the new `domain_items` property of the `billing-history` endpoint.

If there is only one domain-related product for a given domain, it is shown as-is.
If a group contains a Private Registration, it will be called "Domain Registration with Privacy Protection"; otherwise just "Domain Registration".

| Before | After |
| --- | --- |
| <img width="671" alt="screen shot 2018-03-06 at 19 10 46" src="https://user-images.githubusercontent.com/2070010/37052772-2aef9406-2172-11e8-829e-178ca1f355d4.png"> | <img width="675" alt="screen shot 2018-03-15 at 14 53 36" src="https://user-images.githubusercontent.com/2070010/37470992-b3e09f54-2860-11e8-8553-2cd462fb2287.png"> |

### Notes

- Please note that this PR does not change the email receipt, which has a different format and is generated in a different manner:

<img width="554" alt="screen shot 2018-03-06 at 19 21 27" src="https://user-images.githubusercontent.com/2070010/37053382-e8214b04-2173-11e8-847e-e634a00e0ed8.png">

## Testing instructions

Sandbox the API and apply 10703-code.
Navigate to `/me/purchases/billing`.

- Make sure that items previously marked as "Multiple items" because containing only multiple products for the same domain, are now listed as "Domain Registration" or "Domain Registration with Privacy Protection".
- Items that contain both multiple products for the same domain and other products (e.g. a plan), should still be listed as "Multiple items".

- Open a receipt containing multiple products for the same domain (e.g. domain registration, mapping, private registration).
  - Make sure that the receipt only shows one "Domain Registration" for that domain, and its amount is the sum of all single domain-related products.
  - If the receipt contains both a domain and privacy registration for the same domain, the product shown is "Domain Registration with Privacy Protection".
- Open a receipt containing only one product for a domain (e.g. only domain mapping).
  - Make sure that the receipt shows "Domain Mapping".

kwight commented:
This works well for me; I did have one possibility for simplification, which I've left on D10703-code.

vindl commented on client/me/billing-history/receipt.jsx:
```diff
@@ -142,7 +143,13 @@ class BillingReceipt extends React.Component {
-		const items = transaction.items.map( item => {
+
+		const groupedTransactionItems = [
+			...reject( transaction.items, { product: 'Domain' } ),
```
Maybe we could create a helper for this to avoid repeating it bellow. Also adding a comment explaining why are we removing the domain products might be helpful, because it might not be obvious to someone reading this without proper context.

Copons commented:
@vindl Sorry for the very bad timing, but I've just pushed a major change on the logic here.

The previous approach relied entirely on the backend because somehow I missed that we had a format currency lib in Calypso as well.
Now, instead, it's plenty enough to add the raw amount and the currency to the response, and build the list here in the front end.

The behaviour is (should be!) the same as before, just now it's almost completely handled by Calypso.

gwwar commented:
@Automattic/payments did you have a preference here for moving more of the logic to the endpoint or let the Calypso client deal with it?

gwwar commented on client/me/billing-history/transactions-table.jsx:
```diff
@@ -105,19 +105,20 @@ class TransactionsTable extends React.Component {
 		}
 
-		return name;
+		const groupedTransactionItems = groupDomainProducts( transaction.items, this.props.translate );
```
Can `groupedTransactionItems` ever be an empty array?

Copons replied:
Yes, if the receipt has no items - which I'd say it'd be quite a corner case.

gwwar commented on client/me/billing-history/transactions-table.jsx:
```diff
@@ -105,19 +105,20 @@ class TransactionsTable extends React.Component {
+		}
+
+		return this.serviceNameDescription( {
+			...groupedTransactionItems[ 0 ],
```
Let's avoid the magic number here,  either a variable name would be helpful here, or we can use an array destructure like so:
```jsx
const [ transactionItem, ...groupedItems ] = groupDomainProducts( transaction.items, this.props.translate );
if ( groupedItems.length > 0 ) {
    // multiple items
}

return this.serviceNameDescription( {
    ...transactionItem,
    plan: capitalPDangit( titleCase( transactionItem.variation ) ),
} );
```

Copons replied:
It's not a magic number, but the actual first element of a 1-element array. 🙂 

The idea is that, if the (grouped) receipt contains only 1 item (of any type), we show that item's details in the receipts list.

On the other hand, if the receipt contains multiple items (again, of any type), we show "Multiple items".
In this case, we could have, for example:
- 1 "normal" item (e.g. "WP.com Biz subscription") + 1 domain item
- 2 normal items + 0 domain items
- 0 normal items + 2 domain items

So, in your example, `[ transactionItem, ...groupedItems ]` could end up with a semantically wrong destructuring.
For example, if there are 2 `transactionItems`, the second would end up into the `groupedItems` array, which should only contain domain groups.

The result would be the same, but in my opinion using the first and only element of a length-1 array should be clear enough. 🙂

gwwar replied:
>  first element of a 1-element array

For readability a variable name is still very helpful! When folks are passing back arrays of misc values not having to dig through the generating function is great.

gwwar commented:
@designsimply I think keeping private domain registration separate still makes sense. What do folks think? 

There aren't any blockers on current code, so let's clear up this point before 🚢

designsimply commented:
As a customer, I don't need to see them listed separately but I would like to see what was actually purchased reflected in the name itself (i.e. "Domain Registration" or "Domain Registration with Privacy Protection"). 

This applies to billing history as well as when managing a purchase:

![screen shot 2018-03-09 at fri mar 9 1 19 48 pm](https://user-images.githubusercontent.com/1119271/37228092-ae45edf6-239c-11e8-8aa5-06d92ef41c27.png)
<sup>Seen at https://calypso.live/me/purchases/lovewhatyoudo.blog/7755174?branch=fix/2525-billing-receipt-group-domain-products (with D10703 and the API sandboxed) using Firefox 58.0.2 on macOS 10.13.3.</sup>

designsimply commented:
Nitpick: I noticed that a refunded domain shows the free *.wordpress.com version of the domain and not the domain name that was refunded. Is this the expected behavior?

![screen shot 2018-03-09 at fri mar 9 1 26 37 pm](https://user-images.githubusercontent.com/1119271/37228624-7f33dbd4-239e-11e8-8bdb-87b50762d667.png)
<sup>Seen at https://calypso.live/me/purchases/lovewhatyoudo.blog/7755174?branch=fix/2525-billing-receipt-group-domain-products (with D10703 and the API sandboxed) using Firefox 58.0.2 on macOS 10.13.3.</sup>

![screen shot 2018-03-09 at fri mar 9 1 31 27 pm](https://user-images.githubusercontent.com/1119271/37228632-827c600e-239e-11e8-98cb-9ae47e59ee8f.png)
<sup>Seen at https://calypso.live/me/purchases/billing/22766598?branch=fix/2525-billing-receipt-group-domain-products (with D10703 and the API sandboxed) using Firefox 58.0.2 on macOS 10.13.3.</sup>

Copons commented:
@gwwar I've updated the array following your suggestion, but with a more generic-but-semantically-correct naming: `[ transactionItem, ...moreTransactionItems ]`.

@gwwar @designsimply I've updated the utility function to handle domains with private registration.
Now they are reported as `Domain Registration with Privacy Protection` instead of simply `Domain Registration`.
**Note**: this requires an update of D10703-code.

This update also fixes a very dumb l10n bug: before, the check on domain-related products was done on the pre-translated product name (`Domain`), so it could pretty much only work in English.
Now, the endpoint also returns the product slug (`wp-domains`) which is a more appropriate string to check.

@designsimply I'm not exactly sure how to test the refund.
I've tried to buy (with my card, because AFAIK credits purchases can't be refunded 🤔) a domain and refund it immediately after but this is what I've got in the domain field:

<img width="733" alt="screen shot 2018-03-12 at 12 19 23" src="https://user-images.githubusercontent.com/2070010/37283409-c52ce9a2-25ef-11e8-8b84-f6ec5d046f18.png">

kwight commented:
Noting that the latest version of this PR and d10703-code work well for me, no issues. I'm also seeing the broken logo that @designsimply mentioned.

I think we just need a Cobalt signoff on this and d10703-code.

Copons commented:
Pinging @Automattic/cobalt for a last look! 🙇 

@kwight @designsimply I think the broken logo only happens when sandboxing? 🤔

prettyboymp commented:
@Copons  I'm seeing receipts for `Domain Mapping` only show as `Domain Registration`.  If there is only one item for the domain and it's mapping, it should show as `Domain Mapping` for the user.

Copons commented:
@prettyboymp Great catch, I somehow thought domain mappings were reported as registration+mapping anyway.

In d2c5043 I've added a check to the groups. If they contain only 1 item, they're returned without changes.
In other words, receipts containing only 1 (for the same domain) between "Domain Registration", "Domain Mapping", "Private Registration", will not be grouped.

kwight commented:
> I think the broken logo only happens when sandboxing?

It's missing for me on prod also, which makes sense; the problem is this URL that is 404: `https://s2.wp.com/wp-content/themes/a8c/public-api/tabs/images/wpcom-logo-420.png`. That `public-api` is the result of a `get_stylesheet_directory_uri()` call in one of the billing history classes server-side; I'll do a separate PR to fix.
